lib/cluster/cluster.c

```diff
@@ -119,6 +119,22 @@ crm_cluster_disconnect(crm_cluster_t *cluster)
     }
 }
 
+/*!
+ * \brief Free a \p crm_cluster_t object and its dynamically allocated members
+ *
+ * \param[in,out] cluster  Cluster object to free
+ */
+void
+crm_cluster_free(crm_cluster_t *cluster)
```

Comment:
Should this be `pcmk_cluster_free()` (or something else like `pcmk_free_crm_cluster()`? On the one hand, `pcmk` is the preferred prefix. On the other, the data type being freed is `crm_*`.

Reply:
What does the constructor need to do as of right now? There are three `calloc()` locations.
* `pacemaker-fenced.c` sets `destroy`, `cpg.cpg_deliver_fn`, and `cpg.cpg_confchg_fn` if `#SUPPORT_COROSYNC`.
* `attrd_corosync.c` does the same unconditionally (based on the name it's probably not compiled if we don't support corosync)
* `controld_control.c` just `calloc()`s and lets `crm_connect_corosync()` initialize the members.

Should the initializer just `calloc()`, assert if `NULL`, and return?

I need to learn more about ABI compatibility and such, because as of right now I don't really understand how a constructor returning the object makes adding members to the end of the struct definition safer.

Reply:
Yes, the constructor should just calloc() and assert.

ABI compatibility is about code compiled with earlier versions of headers still working when linked against a newer version of the library. If we have a `struct s { int a; }`, then if external code does either `struct s s = { 0, };` or `struct s *s = calloc(1, sizeof(s));`, then adding an `int b;` member at the end will break that code, because either way the object only has enough memory to hold `a` (it was determined at the external code's compile-time). However if we define a constructor that does the calloc(), and the external code uses that, then it'll always get the current size of the struct (it was determined at the library's compile-time).

Reply:
The way we typically do it, with a public header definition and a constructor, is safer but still easily misused since we can only ask external developers to use the constructor -- they could still use one of the problematic forms. The safest way is to use opaque objects, where only a typedef for the struct pointer is in the public header, along with accessor functions. However it's less convenient for both us and the external developers (having to write/use accessor functions vs direct struct member access).